daemon/emer-daemon.c

```diff
@@ -548,6 +566,89 @@ add_events_to_builders (GVariant       **events,
     }
 }
 
+static gboolean
+parse_event_id (const gchar *unparsed_event_id,
+                uuid_t       parsed_event_id)
+{
+  gint parse_failed = uuid_parse (unparsed_event_id, parsed_event_id);
+  if (parse_failed != 0)
+    {
+      g_warning ("Attempt to parse UUID \"%s\" failed. Make sure you created "
+                 "this UUID with uuidgen -r. You may need to sudo apt-get "
+                 "install uuid-runtime first.", unparsed_event_id);
+      return FALSE;
+    }
+
+  return TRUE;
+}
+
+static gboolean
+report_invalid_data_in_cache_on_idle (CacheMetricEventData *callback_data)
+{
+  EmerDaemon *self = callback_data->daemon;
+  const char *event_id = callback_data->event_id;
+  GVariant *payload = callback_data->payload;
+  GVariant *actual_payload = NULL;
+
+  gint64 relative_time;
+  if (!emtr_util_get_current_time (CLOCK_BOOTTIME, &relative_time))
+    {
+      g_critical ("Getting relative timestamp failed.");
+      goto free;
+    }
+
+  uuid_t parsed_event_id;
+  if (!parse_event_id (event_id, parsed_event_id))
+    {
+      g_critical ("Could not parse event ID");
+      goto free;
+    }
+
+  GVariantBuilder uuid_builder;
+  get_uuid_builder (parsed_event_id, &uuid_builder);
+  GVariant *event_id_variant = g_variant_builder_end (&uuid_builder);
+
+  if (payload != NULL)
+    {
+      actual_payload = payload;
+    }
+  else
+    {
+      GVariant *unboxed_variant = g_variant_new_boolean (FALSE);
+      GVariant *empty_payload = g_variant_new_variant (unboxed_variant);
+      actual_payload = g_variant_ref_sink (empty_payload);
+    }
+
+  emer_daemon_record_singular_event (self,
+                                     getuid (),
+                                     event_id_variant,
+                                     relative_time,
+                                     payload != NULL,
+                                     actual_payload);
+ free:
+  if (actual_payload != NULL)
+    g_variant_unref (actual_payload);
+  g_free (callback_data);
+
+  return G_SOURCE_REMOVE;
+}
+
+static void
+report_invalid_data_in_cache (EmerDaemon *self,
+                              const char *event_id,
+                              GVariant   *payload)
+{
+  CacheMetricEventData *callback_data = g_new (CacheMetricEventData, 1);
+  callback_data->daemon = self;
+  callback_data->event_id = event_id;
+  callback_data->payload = payload != NULL ? g_variant_ref (payload) : NULL;
```

Comment:
I don't think you need to ref this, unless you want to use the variant after the idle.